extensions/INTEL/INTEL_shader_integer_functions2.txt

```diff
@@ -0,0 +1,268 @@
+Name
+
+    INTEL_shader_integer_functions2
+
+Name Strings
+
+    GL_INTEL_shader_integer_functions2
+
+Contact
+
+    Ian Romanick <ian.d.romanick@intel.com>
+
+Contributors
+
+
+Status
+
+    In progress
+
+Version
+
+    Last Modification Date: 11/14/2019
+    Revision: 4
+
+Number
+
+    OpenGL Extension #547
+    OpenGL ES Extension #323
+
+Dependencies
+
+    This extension is written against the OpenGL 4.6 (Core Profile)
+    Specification.
+
+    This extension is written against Version 4.60 (Revision 03) of the OpenGL
+    Shading Language Specification.
+
+    GLSL 1.30 (OpenGL), GLSL ES 3.00 (OpenGL ES), or EXT_gpu_shader4 (OpenGL)
+    is required.
+
+    This extension interacts with ARB_gpu_shader_int64.
+
+    This extension interacts with AMD_gpu_shader_int16.
+
+    This extension interacts with OpenGL 4.6 and ARB_gl_spirv.
+
+    This extension interacts with EXT_shader_explicit_arithmetic_types.
+
+Overview
+
+    OpenCL and other GPU programming environments provides a number of useful
+    functions operating on integer data.  Many of these functions are
+    supported by specialized instructions various GPUs.  Correct GLSL
+    implementations for some of these functions are non-trivial.  Recognizing
+    open-coded versions of these functions is often impractical.  As a result,
+    potential performance improvements go unrealized.
+
+    This extension makes available a number of functions that have specialized
+    instruction support on Intel GPUs.
+
+New Procedures and Functions
+
+    None
+
+New Tokens
+
+    None
+
+IP Status
+
+    No known IP claims.
+
+Modifications to the OpenGL Shading Language Specification, Version 4.60
+
+    Including the following line in a shader can be used to control the
+    language features described in this extension:
+
+      #extension GL_INTEL_shader_integer_functions2 : <behavior>
+
+    where <behavior> is as specified in section 3.3.
+
+    New preprocessor #defines are added to the OpenGL Shading Language:
+
+      #define GL_INTEL_shader_integer_functions2        1
+
+Additions to Chapter 8 of the OpenGL Shading Language Specification
+(Built-in Functions)
+
+    Modify Section 8.8, Integer Functions
+
+    (add a new rows after the existing "findMSB" table row, p. 161)
+
+    genUType countLeadingZeros(genUType value)
+
+    Returns the number of leading 0-bits, stating at the most significant bit,
+    in the binary representation of value.  If value is zero, the size in bits
+    of the type of value or component type of value, if value is a vector will
```

Comment:
In the SPIR-V spec we tweaked the "if value is a vector" to use parenthesis for clarity.

> If the value is zero, the size in bits of the type of value or component type of value (if value is a vector) will be returned.

Reply:
Fixed locally.